std/algorithm/iteration.d

```diff
@@ -4289,3 +4289,478 @@ unittest
+    assert(iota(3).combinations(2).array == [[0, 1], [0, 2], [1, 2]]);
+    assert("AB"d.combinations(2).array == ["AB"d]);
+    assert("ABC"d.combinations(2).array == ["AB"d, "AC"d, "BC"d]);
+}
+
+pure nothrow @safe unittest
+{
+    import std.array: array;
+    import std.range: iota, dropOne;
+    assert(iota(0).combinations.array == []);
+    assert(iota(2).combinations.array == [[0], [1]]);
+    assert("ABCD"d.combinations(2).array == ["AB"d, "AC"d, "AD"d, "BC"d, "BD"d, "CD"d]);
+    assert(iota(2).array.combinations.front == [0]);
+
+    // is copyable?
+    auto a = iota(2).combinations;
+    assert(a.front == [0]);
+    assert(a.save.dropOne.front == [1]);
+    assert(a.front == [0]);
+
+    // test larger combinations
+    auto expected5 = [[0, 1, 2], [0, 1, 3], [0, 1, 4],
+                      [0, 2, 3], [0, 2, 4], [0, 3, 4],
+                      [1, 2, 3], [1, 2, 4], [1, 3, 4],
+                      [2, 3, 4]];
+    assert(iota(5).array.combinations(3).array == expected5);
+    assert(iota(4).array.combinations(3).array == [[0, 1, 2], [0, 1, 3], [0, 2, 3], [1, 2, 3]]);
+    assert(iota(3).array.combinations(3).array == [[0, 1, 2]]);
+    assert(iota(2).array.combinations(3).array == []);
+    assert(iota(1).array.combinations(3).array == []);
+
+    assert(iota(3).array.combinations(2).array == [[0, 1], [0, 2], [1, 2]]);
+    assert(iota(2).array.combinations(2).array == [[0, 1]]);
+    assert(iota(1).array.combinations(2).array == []);
+
+    assert(iota(1).array.combinations(1).array == [[0]]);
+}
+
+/**
+Lazily computes the all k-combinations of $(D r) with repetitions.
+A k-combination with repetitions, or k-multicombination, or multisubset of size k from a set S is given by a sequence of k not necessarily distinct elements of S, where order is not taken into account.
+Imagine this as the product filtered for only ordered items.
+
+For example $(D"AB".combinationsRepeat(2).array) returns $(D["AA", "AB", "BB"]).
+
+Params:
+    r = RandomAccessRange or string as origin
+    k = number of combinations
+
+Returns:
+    Forward range which yields the k-multicombinations items
+*/
+auto combinationsRepeat(Range)(Range r, size_t k=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
+    assert(k >= 1, "Invalid k");
+}
+body
+{
+    import std.array: array;
+    import std.range: indexed;
+    static struct CombinationRepeat
+    {
+        ElementType!Range[] front;
+        private Range _r;
+        private size_t[] _state;
+
+        private size_t _repeat, _length;
+        private size_t _max_states, _pos;
+        private bool _lengthComputed;
+
+        bool empty;
+
+        this(Range r, size_t repeat) pure nothrow @safe
+        {
+            _r = r;
+            _length = r.length;
+            _repeat = repeat;
+
+            // set initial state
+            _max_states = length;
+            if (_length > 0 && _max_states > 0)
+            {
+                _state = new size_t[repeat];
```

Comment:
I do admit this is very confusing - the number of elements (`r.length`) is cached (in the documentation I read that it is _not_ guaranteed to be constant) and it is used quite often (in the `while` loop of `popFront`. 
Moreover `max_states` computes the all possible states as this makes computations a lot easier and provides the user with a nice `.length` property. 
To avoid this confusion I also renamed `lengthComputed` to `isNrStatesComputed`.